src/sync/cache.rs

```diff
@@ -0,0 +1,209 @@
+use color_eyre::Help;
+use color_eyre::{eyre::WrapErr, Result};
+use std::io::{BufReader, BufWriter};
+use std::time::{Duration, SystemTime};
+use std::{fs, net::IpAddr};
+
+use serde::{Deserialize, Serialize};
+
+const EXPIRATION: Duration = Duration::from_secs(60 * 60 * 24 * 7 * 4 * 2);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Entry {
+    ip: IpAddr,
+    last_updated: SystemTime,
+}
+
+#[derive(Debug)]
+/// route cache that might contain outdated entries
+/// that should be removed
+pub struct Cached(Vec<Entry>);
+
+impl Cached {
+    pub fn load() -> Result<Self> {
+        let f = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open("routes.json")?;
+
+        if f.metadata()?.len() == 0 {
+            return Ok(Cached(Vec::new()));
+        }
+
+        let r = BufReader::new(f);
+        let entries = serde_json::from_reader(r)
+            .wrap_err("could not parse adress in file")
+            .suggestion("try removing `routes.json`")
+            .note("if that does not work please open an issue")?;
+        Ok(Cached(entries))
+    }
+
+    #[must_use]
+    fn n_recent(&self) -> usize {
+        self.0
+            .iter()
+            .filter(|e| e.last_updated.elapsed().unwrap_or(Duration::ZERO) < EXPIRATION)
+            .count()
+    }
+
+    #[must_use]
+    pub fn blocked_ips(self) -> Vec<IpAddr> {
+        self.0.into_iter().map(|e| e.ip).collect()
+    }
+
+    #[must_use]
+    pub fn update(mut self, new: Vec<IpAddr>) -> Option<Routes> {
+        self.0.extend(new.into_iter().map(|ip| Entry {
+            ip,
+            last_updated: SystemTime::now(),
+        }));
+        dedup_keep_newest(&mut self.0);
+
+        if self.0.is_empty() {
+            return None;
+        }
+
+        if self.n_recent() < 2 {
+            return Some(Routes(self.0));
+        }
+
+        /* TODO: replace with drain_filter once stabelized <16-02-23, dvdsk> */
+        let mut i = 0;
+        while i < self.0.len() {
+            if let Ok(age) = self.0[i].last_updated.elapsed() {
+                if age > EXPIRATION {
+                    self.0.remove(i);
+                    continue;
+                }
+            }
+            i += 1;
+        }
+
+        Some(Routes(self.0))
+    }
+}
+
+pub struct Routes(Vec<Entry>);
+
+impl Routes {
```

Comment:
Name doesn't make clear what the difference between Routes and Cached is. Suggestion: UpdatedRoutes?